### PR TITLE
Add LVI mitigation for OpenSSL 3

### PR DESCRIPTION
**Pull request by justanotherminh**
This PR copies and modifies LVI-mitigated OpenSSL 3 assembly files with control flow from Intel.

### CODE REVIEW

File: 3rdparty/openssl/intel-sgx-ssl-3
Comment (dcarpente) [inline 3rdparty/openssl/intel-sgx-ssl-3:1]: Can we point to a commit tagged as release by Intel SSL? E.g. I think we should use this commit ef50655:
https://github.com/intel/intel-sgx-ssl/releases/tag/3.0_Rev1

File: 3rdparty/openssl/CMakeLists.txt [hunk 1]
```diff
@@ -76,7 +78,10 @@ if (OE_SGX)
   # Special case where assembly file is generated by different script
   # (crypto/sha/asm/sha256-mb-x86_64.pl)
   list(APPEND OPENSSL_GENERATED_BYPRODUCTS crypto/sha/asm/sha256-x86_64)
-  list(APPEND OPENSSL_3_GENERATED_BYPRODUCTS crypto/sha/asm/sha256-x86_64)
+  # list(APPEND OPENSSL_3_GENERATED_BYPRODUCTS crypto/sha/asm/sha256-x86_64)
```
Comment (dcarpente) [inline 3rdparty/openssl/CMakeLists.txt:81]: Remove commented-out line

File: 3rdparty/openssl/CMakeLists.txt [hunk 2]
```diff
@@ -76,7 +78,10 @@ if (OE_SGX)
...
+
+  list(APPEND OPENSSL_3_GENERATED_BYPRODUCTS crypto/bn/asm/rsaz-2k-avx512
```
Comment (dcarpente) [inline 3rdparty/openssl/CMakeLists.txt:83]: Add a comment explaining why these files are appended seperately. (E.g. "Intel SSL repo doesn't include these assembly files, we need to generate them using OpenSSL .pl scripts as we do for OpenSSL 1.1.1")

File: 3rdparty/openssl/CMakeLists.txt [hunk 3]
```diff
@@ -354,16 +359,51 @@ macro (append_generate_command LIST FILE_NAME SOURCE_DIR DEST_DIR)
     ${DEST_DIR}/${FILE_NAME}.s)
 endmacro ()
 
+macro (append_generate_3_command LIST FILE_NAME SOURCE_DIR DEST_DIR)
```
Comment (dcarpente) [inline 3rdparty/openssl/CMakeLists.txt:362]: Maybe we rename this and the above macro, as this is no longer really generating a file. This could be called "append_copy_assembly_command" and the above macro could be "append_generate_assembly_command"

File: 3rdparty/openssl/CMakeLists.txt [hunk 4]
```diff
@@ -431,33 +458,15 @@ ExternalProject_Add(
...
-  INSTALL_COMMAND "")
-
 add_dependencies(openssl_includes openssl_configure)
 add_dependencies(openssl_copied openssl_configure)
 add_dependencies(openssl_generated openssl_includes)
 add_dependencies(openssl_generated openssl_copied)
 
 add_dependencies(openssl_3_includes openssl_3_configure)
-add_dependencies(openssl_3_copied openssl_3_configure)
+# add_dependencies(openssl_3_copied openssl_3_configure)
```
Comment (dcarpente) [inline 3rdparty/openssl/CMakeLists.txt:467]: Remove both commented-out lines here

File: 3rdparty/openssl/CMakeLists.txt [hunk 5]
```diff
@@ -1352,9 +1361,12 @@ add_enclave_library(
   opensslcrypto STATIC ${OPENSSL_1_CRYPTO_SOURCE}
   ${OPENSSL_GENERATED_BYPRODUCTS} ${OPENSSL_COPIED_BYPRODUCTS})
 
-add_enclave_library(
-  opensslcrypto_3 STATIC ${OPENSSL_3_CRYPTO_SOURCE}
-  ${OPENSSL_3_GENERATED_BYPRODUCTS} ${OPENSSL_3_COPIED_BYPRODUCTS})
+# add_enclave_library(
```
Comment (dcarpente) [inline 3rdparty/openssl/CMakeLists.txt:1364]: Remove commented out line

**Review by dcarpente — COMMENTED** [review]
Left some comments.

**Comment by justanotherminh** [comment]
bors try

**Comment by CyanDevs** [comment]
bors try

File: 3rdparty/openssl/CMakeLists.txt [hunk 6]
```diff
@@ -63,11 +64,12 @@ list(
   APPEND
   OPENSSL_3_ASSEMBLY_SCRIPTS
   ${OPENSSL_ASSEMBLY_SCRIPTS}
+  crypto/x86_64cpuid
+  crypto/aes/asm/aesni-x86_64
+  crypto/aes/asm/vpaes-x86_64
+  crypto/sha/asm/sha256-x86_64
   crypto/aes/asm/aes-x86_64
-  crypto/aes/asm/bsaes-x86_64
```
Comment (radhikaj) [inline 3rdparty/openssl/CMakeLists.txt:67]: why were these removed

Comment (justanotherminh) [reply 3rdparty/openssl/CMakeLists.txt:67]: They were added back later in the script. I'm using this file list exclusively for assembly files that have been processed by Intel. These are not.

File: 3rdparty/openssl/remove_lfence.pl [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#! /usr/bin/env perl
```
[the remark this answers is not in the document]
Comment (justanotherminh) [reply 3rdparty/openssl/remove_lfence.pl:1]: This file removes every lfence line that has a comment saying "load_only" in the assembly scripts provided by Intel for control flow LVI mitigation. If you check the bash install script in intel-sgx-ssl you'll see that it's what they do.

[the remark this answers is not in the document]
Comment (dcarpente) [reply 3rdparty/openssl/remove_lfence.pl:1]: Can you add a comment at the top of this file explaining what it does?

**Review by dcarpente — APPROVED** [review]
Thank you, just make sure to squash before submitting!

**Review by mingweishih — APPROVED** [review]

**Review by dcarpente — APPROVED** [review]

**Review by CyanDevs — APPROVED** [review]
